include/envoy/api/io_error.h

```diff
@@ -69,6 +40,9 @@ using IoErrorPtr = std::unique_ptr<IoError, IoErrorDeleterType>;
 /**
  * Basic type for return result which has a return code and error code defined
  * according to different implementations.
+ * If the call succeeds, |err_| is nullptr and |rc_| is valid. Otherwise |err_|
+ * can be passed into IoError::getErrorCode() to extract the error. In this
+ * case, |rc_| is invalid.
```

Comment:
given this interface can you add method:
```
 bool ok() const { return err_ == nullptr; }
```

I had to also read below at usages to understand that rc is really the interface-dependent return value. Can you call it return_value_ instead of rc_ (which sounds like it might be errno or something), and use ReturnValue rather than T as your template param?

Reply:
and of course use that method at call-sites rather than directly dereferencing err_.

Reply:
> given this interface can you add method:
> 
> ```
>  bool ok() const { return err_ == nullptr; }
> ```
> I had to also read below at usages to understand that rc is really the interface-dependent return value. Can you call it return_value_ instead of rc_ (which sounds like it might be errno or something), and use ReturnValue rather than T as your template param?

This renaming will touch even larger scope of files which is not related to socket read/write, since IoError is used in filesystem now. I would prefer do it separately so that this PR doesn't looks confusing.

Reply:
Can we add the ok() method now and change usage in files already in this PR, and follow up with the file-system changes later?

Reply:
Adding ok() sounds good to me

Reply:
done